DOC: add missing details to linalg.lstsq docstring

nilokr opened the pull request:
It turns out that `linalg.lstsq` also minimizes the 2-norm of `x` when `a` is rank-deficient. I found that by searching [the documentation for the LAPACK library](https://www.netlib.org/lapack/lug/node27.html), which is the current implementation of `lstsq`.

This PR rewrites the docstring for `linalg.lstsq` such that it contains details for all cases.

charris commented on numpy/linalg/linalg.py:
```diff
@@ -2171,13 +2171,13 @@ def lstsq(a, b, rcond="warn"):
+    ``a @ x = b``, where `a` is an `m`-by-`n` matrix and `b` is a `m` element 
+    vector. When the system is overdetermined (``m >= n`` and ``rank(a) = n``), 
+    a solution `x` that minimizes the Euclidean 2-norm ``||b-ax||`` is found.
+    Else, when the system is underdetermined (``m < n`` and ``rank(A) = m``) 
```
Maybe "If there are multiple minimizing solutions, the one with the smallest norm is returned". Note that the norm of the solution depends on column scaling (units) and choice of variables, so the smallest norm solution is seldom meaningful when the solution is not unique.

krnilo replied:
Sounds better - I've updated my commit to reflect the suggestion, thanks!

The reason I stumbled on this smallest norm matter was that I was specifically looking for that solution in a rank-deficient problem. I was intrigued I got exactly what I wanted for free, so I did some digging and found why :)

nilokr committed:
DOC: add missing details to linalg.lstsq docstring

It turns out that lstsq also minimizes the 2-norm of x when a is
rank-deficient. I found that by searching the documentation for the
LAPACK library, which is the current implementation of lstsq (as of Dec
2020). Ref: https://www.netlib.org/lapack/lug/node27.html

charris commented on numpy/linalg/linalg.py:
```diff
@@ -2171,13 +2171,14 @@ def lstsq(a, b, rcond="warn"):
     If `a` is square and of full rank, then `x` (but for round-off error)
     is the "exact" solution of the equation. Else, `x` minimizes the
-    Euclidean 2-norm :math:`|| b - a x ||`.
+    Euclidean 2-norm :math:`||b-ax||`. If there are multiple minimizing 
```
Keep spaces around `-` for readability.

charris replied:
I fixed that.

charris committed:
STY: Add spaces around '-'.

[ci skip]

charris commented:
Thanks @krnilo .

matthew-brett commented:
Worth pointing out that this is the behavior of the pseudoinverse?
